Cloud Native Chapter 6

ardalis opened the pull request:
## Summary

Chapter 6 md files and images

ardalis committed:
Adding ch6 files

EugeneKramer commented on docs/architecture/cloud-native/resilient-communications.md:
```diff
@@ -0,0 +1,70 @@
+
+Throughout this book, we've evangelized the merits of moving beyond traditional monolithic application design and embracing a microservice-based architecture where a set of distributed, self-contained services run independently and communicate with each other using standard communication protocols such as HTTP and HTTPS. While such an architecture buys you many important benefits, it also present ...
+
+- *Out-of-process network communication* Each service communicates over a network protocol which introduces network congestion, latency and transient faults 
```
```suggestion
- *Out-of-process network communication* - Each service communicates over a network protocol which introduces network congestion, latency and transient faults.
```

EugeneKramer commented on docs/architecture/cloud-native/resilient-communications.md:
```diff
@@ -0,0 +1,70 @@
+
+- *Out-of-process network communication* Each service communicates over a network protocol which introduces network congestion, latency and transient faults 
+- *Service discovery* - With each service running across a cluster of machines with its own IP address and port, how do services discover and communicate with each other?
+- *Resiliency and failover* How do we manage short-lived failures and keep the system stable?
```
```suggestion
- *Resiliency and failover* - How do we manage short-lived failures and keep the system stable?
```

EugeneKramer commented on docs/architecture/cloud-native/resilient-communications.md:
```diff
@@ -0,0 +1,70 @@
+- *Out-of-process network communication* Each service communicates over a network protocol which introduces network congestion, latency and transient faults 
+- *Service discovery* - With each service running across a cluster of machines with its own IP address and port, how do services discover and communicate with each other?
+- *Resiliency and failover* How do we manage short-lived failures and keep the system stable?
+- *Load balancing* How does inbound traffic get distributed across multiple instances of a service?
```
```suggestion
- *Load balancing* - How does inbound traffic get distributed across multiple instances of a service?
```

EugeneKramer commented on docs/architecture/cloud-native/resilient-communications.md:
```diff
@@ -0,0 +1,70 @@
+- *Service discovery* - With each service running across a cluster of machines with its own IP address and port, how do services discover and communicate with each other?
+- *Resiliency and failover* How do we manage short-lived failures and keep the system stable?
+- *Load balancing* How does inbound traffic get distributed across multiple instances of a service?
+- *Security* How are security concerns such as transport-level encryption and certificate management enforced?
```
```suggestion
- *Security* - How are security concerns such as transport-level encryption and certificate management enforced?
```

EugeneKramer commented on docs/architecture/cloud-native/resilient-communications.md:
```diff
@@ -0,0 +1,70 @@
+- *Resiliency and failover* How do we manage short-lived failures and keep the system stable?
+- *Load balancing* How does inbound traffic get distributed across multiple instances of a service?
+- *Security* How are security concerns such as transport-level encryption and certificate management enforced?
+- *Distributed Monitoring* How do you correlate and capture traceability and monitoring for a single request across multiple consuming services?
```
```suggestion
- *Distributed Monitoring* - How do you correlate and capture traceability and monitoring for a single request across multiple consuming services?
```

ardalis committed:
Apply suggestions from code review

Thanks! There were unicode long dashes there before that were removed instead of replaced with -.

Co-Authored-By: Eugene Kramer <EugeneKramer@users.noreply.github.com>

mairaw commented on docs/architecture/cloud-native/application-resiliency-patterns.md:
```diff
@@ -0,0 +1,75 @@
+
+**Figure 6-2**. Polly Resiliency Framework Features
+
+Note in the figure above how the various resiliency policies apply to operation requests, whether coming from an external client or another backend service. The goal is to correctly compensate the request for a that be might intermittently unavailable for a short time duration (i.e., in seconds). These interruptions typically manifest themselves with specific HTTP status codes as shown in Figure
```
for a that -> not sure what word you wanted to use here

mairaw commented on docs/architecture/cloud-native/application-resiliency-patterns.md:
```diff
@@ -0,0 +1,75 @@
+
+Question: Would you retry an HTTP Status Code of 403 - Forbidden? No. In that case, the system is functioning properly and telling you are simply not authorized to perform the operation that you are requesting. Care must be taken in selecting appropriate operations to retry.
+
+As recommended in Chapter 1, Microsoft developers targeting cloud native applications should be building on the .NET Core framework. At the time of this writing, version 2.2 is the current release. Version 2.1 introduced the HTTPClientFactory library as a for sending HTTP requests and receiving HTTP responses from a resource identified as a URL. Superseding the original HTTPClient class, the fact ...
```
as a for sending -> as a way for sending?

mairaw committed:
edits

mairaw committed:
fix headings case

mairaw commented on docs/architecture/cloud-native/infrastructure-resiliency-azure.md:
```diff
@@ -0,0 +1,87 @@
+
+**Figure 6-8.** Scale Up vs. Scale Out
+
+- *Scale proportionally.* When scaling a service, think in terms of *resource sets*. If you were to dramatically scale out a specific service, what impact would that have on backend data stores, caches and dependent services? Some resources such as Cosmos DB can scale out proportionally, while many other    cannot. You want to ensure that you do not scale out a resource that will to a point where ...
```
will to a -> will what?

mairaw commented on docs/architecture/cloud-native/infrastructure-resiliency-azure.md:
```diff
@@ -0,0 +1,87 @@
+
+- *Take advantage of platform autoscaling features.* Use built-in autoscaling features whenever possible, rather than custom or third-party mechanisms. Use scheduled scaling rules, where possible, to ensure that resources are available without a startup delay, but add reactive autoscaling to the rules, where appropriate, to cope with unexpected changes in demand. For more information, see Autosca ...
+
+- *Scale-up aggressively.* A final practice would be to scale up aggressively so that you can quickly meet immediate spikes in traffic without losing business. And, then scale down (remove unneeded resources) conservatory in order to keep the system stable. A simple way to implement this might be to set the cool down period (the time to wait between scaling operations) for adding resources to 5 m ...
```
conservatory -> conservatively?

mairaw commented on docs/architecture/cloud-native/infrastructure-resiliency-azure.md:
```diff
@@ -0,0 +1,87 @@
+
+- *Avoid affinity.* A best practice is to ensure a node doesn't require local affinity, often referred to as a *sticky session*. A request should be able to route to any instance. If you do need to persist state, it should be saved to a distributed cache, such as Azure Redis cache.
+
+- *Take advantage of platform autoscaling features.* Use built-in autoscaling features whenever possible, rather than custom or third-party mechanisms. Use scheduled scaling rules, where possible, to ensure that resources are available without a startup delay, but add reactive autoscaling to the rules, where appropriate, to cope with unexpected changes in demand. For more information, see Autosca ...
```
missing link

mairaw commented on docs/architecture/cloud-native/infrastructure-resiliency-azure.md:
```diff
@@ -0,0 +1,87 @@
+
+- *Azure Redis Cache.* The Redis StackExchange client uses a connection manager class that includes retries on failed attempts. The number of retries, specific retry policy and wait time are all configurable.
+
+- *Azure Service Bus.* The Service Bus client exposes a RetryPolicy property that can be configured with a back-off interval, retry count and TerminationTimeBuffer which specifies the maximum time an operation can take. The default policy is 9 maximum retry attempts with a 30 second backoff period between attempts.
```
RetryPolicy property or class (https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.retrypolicy?view=azure-dotnet)?

mairaw committed:
edits

mairaw commented on docs/architecture/cloud-native/resiliency.md:
```diff
@@ -0,0 +1,59 @@
+
+It's challenging to design and deploy a distributed cloud-native application. And, it's equally challenging to keep that application running in an environment where failure is certain. Consider what can go wrong:
+
+- Unexpected [Network Latency](https://www.techopedia.com/definition/8553/network-latency) network latency
```
this is repeated. not sure which one you wanna keep.

mairaw committed:
edits

mairaw committed:
fix heading case

mairaw committed:
edits

mairaw commented:
Review is done @ardalis. But there are some pending comments that we should resolve before merging.

mairaw reviewed (changes requested):
Please look at my questions before we can merge

robvet committed:
Ch 6 cleansed including all of Maria's changes.

robvet committed:
Cleansing of Ch 6. All of Maria's updates are included in this commit.

robvet committed:
Merge branch 'cloudnative-ch6' of https://github.com/ardalis/docs-1 into cloudnative-ch6

robvet committed:
setup

robvet committed:
Minor corrections

robvet committed:
Added Summary placeholders

robvet committed:
Formatting updates. Review and PR to Maria.

mairaw committed:
Merge branch 'cloud-native' into cloudnative-ch6

ardalis committed:
Merge remote-tracking branch 'origin/cnch6' into cloudnative-ch6

ardalis commented:
Should be good to merge, @mairaw

mairaw committed:
fix broken links

mairaw committed:
fix broken links

mairaw committed:
fix broken links

mairaw committed:
Merge branch 'cloud-native' into cloudnative-ch6

ardalis commented:
This one says there's 1 change requested but I don't see what it is.

mairaw commented:
We were having some build issues earlier today. Trying again.

mairaw commented on docs/architecture/cloud-native/infrastructure-resiliency-azure.md:
```diff
@@ -0,0 +1,84 @@
+
+- *Azure Redis Cache.* The Redis StackExchange client uses a connection manager class that includes retries on failed attempts. The number of retries, specific retry policy and wait time are all configurable.
+
+- *Azure Service Bus.* The Service Bus client exposes a [RetryPolicy property](https://docs.microsoft.com/dotnet/api/microsoft.servicebus.retrypolicy?view=azure-dotnet) that can be configured with a back-off interval, retry count, and <xref:Microsoft.ServiceBus.RetryExponential.TerminationTimeBuffer>, which specifies the maximum time an operation can take. The default policy is nine maximum retry ...
```
If that's really the API you were talking about, this is a class

```suggestion
- *Azure Service Bus.* The Service Bus client exposes a [RetryPolicy class](xref:Microsoft.ServiceBus.RetryPolicy) that can be configured with a back-off interval, retry count, and <xref:Microsoft.ServiceBus.RetryExponential.TerminationTimeBuffer>, which specifies the maximum time an operation can take. The default policy is nine maximum retry attempts with a 30-second backoff period between attempts.
```

mairaw committed:
edits

mairaw commented:
Edited the last one that had major issues before. One comment left.

ardalis committed:
Update docs/architecture/cloud-native/infrastructure-resiliency-azure.md

Co-Authored-By: Maira Wenzel <mairaw@microsoft.com>

ardalis commented:
Hopefully good to go now.

mairaw committed:
fix broken link

mairaw committed:
Merge branch 'cloud-native' into cloudnative-ch6
